hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/NodeAddressUpdateHandler.java

```diff
@@ -0,0 +1,69 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with this
+ * work for additional information regarding copyright ownership.  The ASF
+ * licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package org.apache.hadoop.hdds.scm.node;
+
+import org.apache.hadoop.hdds.protocol.DatanodeDetails;
+import org.apache.hadoop.hdds.scm.ha.SCMService;
+import org.apache.hadoop.hdds.scm.ha.SCMServiceManager;
+import org.apache.hadoop.hdds.scm.node.states.NodeNotFoundException;
+import org.apache.hadoop.hdds.scm.pipeline.PipelineManager;
+import org.apache.hadoop.hdds.server.events.EventHandler;
+import org.apache.hadoop.hdds.server.events.EventPublisher;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Handles datanode ip or hostname change event.
+ */
+public class NodeAddressUpdateHandler
+        implements EventHandler<DatanodeDetails> {
+  private static final Logger LOG =
+          LoggerFactory.getLogger(NodeAddressUpdateHandler.class);
+
+  private final PipelineManager pipelineManager;
+  private final NodeDecommissionManager decommissionManager;
+  private final SCMServiceManager serviceManager;
+
+  public NodeAddressUpdateHandler(PipelineManager pipelineManager,
+                                  NodeDecommissionManager
+                                               decommissionManager,
+                                  SCMServiceManager serviceManager) {
+    this.pipelineManager = pipelineManager;
+    this.decommissionManager = decommissionManager;
+    this.serviceManager = serviceManager;
+  }
+
+  @Override
+  public void onMessage(DatanodeDetails datanodeDetails,
+                        EventPublisher publisher) {
+    try {
+      LOG.info("Closing stale pipelines for datanode: {}", datanodeDetails);
+      pipelineManager.closeStalePipelines(datanodeDetails);
```

Comment:
> Is closing pipelines necessary even if using datanode hostname instead of IP
The problem is that pipeline is using IP address. The datanode host name config only impact the ratis address. Ideally I think we should use hostname in pipeline. However, pipeline info is kept in persistent, if we we change the code to only support the hostname, then it will be not backward compatible with the pipelines already in the DB. If we decide to support both hostname and IP address, then the logic would be more complex:  we should figure out one IP is corresponding to one hostname so that pipeline will not be created duplicately. Use hostname in pipeline is a big change. We need to consider more about it. Therefore, I think it should be another PR if we decide to make the change.

> If two or three datanodes of some pipeline are restarted, a new pipeline is created as they register, and the first one or two pipelines are then closed very soon.

Is this a question or statement?

Reply:
It's a statement based on behavior I saw while checking kubernetes test results where 3 datanodes are restarted.  See the log above.  Pipeline `98579454-...` is created when datanode-0 registers with new IP.  It is then closed when datanode-1 registers, new pipeline `62cfbe84-...` is created, etc.

But I guess quick restarts are less realistic in non-test environment, so it may be OK for now.

Reply:
Yes, pipeline `56d494b4-...` is before the restart.  This gets closed when first DN re-registers.  My point is that the new pipeline created after this DN registers gets closed right away when the second DN re-registers.

Reply:
Ah, I got what you mean. This is because the restarts are so quick that SCM thinks these 3 datanodes keep alive. So when the first DN re-registers and tries to create the new pipeline, SCM thinks the other two DN are still good, and thus uses the old info (because the other two DNs have not re-registered yet). Same thing happens when the 2nd DN re-registeres, where SCM the 3rd DN is good and uses its old info.

For quick restarts, I do not think there is a good way for SCM to recognize these DN's are dead/restarted. Therefore, we rely on the logic of this PR for waiting ALLOCATED pipelines to be OPEN. In such way, the block allocation can still get the correct pipeline eventually.